### ISSUE TITLE
Undefined index: samesite

### ISSUE THREAD

**Issue by tersmitten** [bryanjhv/slim-session#60]
When running our tests with `PHP 7.2` and version `3.7.0` of this project we get an `Undefined index: samesite` `notice`. This seems to be because `session_get_cookie_params` is not return the `samesite` key and `Cookie` is expecting it in line number `39`. In `PHP 7.2+` the notice is not present. I can make a PR if you want me to.

Something like this could work:

```diff
# git diff
diff --git a/src/SlimSession/Helper.php b/src/SlimSession/Helper.php
index 9ea3469..15c7c0c 100644
--- a/src/SlimSession/Helper.php
+++ b/src/SlimSession/Helper.php
@@ -125,11 +125,16 @@ class Helper implements \ArrayAccess, \Countable, \IteratorAggregate
             session_write_close();
 
             if (ini_get('session.use_cookies')) {
+                $params = session_get_cookie_params();
+                $old = PHP_VERSION_ID < 70300;
+                if ($old) {
+                    $params['samesite'] = '';
+                }
                 Cookie::set(
                     session_name(),
                     '',
                     time() - 4200,
-                    session_get_cookie_params()
+                    $params
                 );
             }
         }
```

or

```diff
# git diff
diff --git a/src/SlimSession/Cookie.php b/src/SlimSession/Cookie.php
index 911bfe9..a2f63e4 100644
--- a/src/SlimSession/Cookie.php
+++ b/src/SlimSession/Cookie.php
@@ -36,7 +36,7 @@ class Cookie
         ];
 
         $new = PHP_VERSION_ID >= 70300;
-        $samesite = $params['samesite'];
+        $samesite = $params['samesite'] ?? '';
```

or

```diff
# git diff
diff --git a/src/SlimSession/Cookie.php b/src/SlimSession/Cookie.php
index 911bfe9..40472d9 100644
--- a/src/SlimSession/Cookie.php
+++ b/src/SlimSession/Cookie.php
@@ -36,15 +36,15 @@ class Cookie
         ];
 
         $new = PHP_VERSION_ID >= 70300;
-        $samesite = $params['samesite'];
         if ($new) {
-            $args['samesite'] = $samesite;
+            $args['samesite'] = $params['samesite'];
             if ($set) {
                 $args = [$set[0], $set[1], $args];
             } else {
                 $args = [$args];
             }
         } else {
+            $samesite = $params['samesite'] ?? '';
```

What do you think?

**Comment by bryanjhv** [comment]
Hi, thank you for the report, didn't notice about that part.

Let's review the solutions you have posted before I suggest another one:

1.  Well, we're required to do change `Cookie.php` because it's where I handle odds for PHP 7.3+.
2.  We can't use null coalesce operator (`??`) introduced in PHP  7, because Slim 3 depends on PHP 5.5+.
3.  Same as (2), but it's repetitive, even if it's a lot clearer.

Having said that... I think it's a good use for the error control operator (`@`) because:

- `$params['samesite']` would return `NULL`, PHP 7.3+ handles it as empty string (`samesite` key must be a string).
- It works on any PHP version, and given all versions (2.x, 3.x, 4.x) for this package need to be fixed, we reuse same code.

Please let me know what do you think.
Would that work? Or is `isset&get` mambo for 2.x and 3.x better, and `??` for 4.x (your 2nd proposal).

**Comment by tersmitten** [comment]
I think that using the `@` operator is a bad practise (to be avoided). But something like this would work in all versions:

```diff
# git diff
diff --git a/src/SlimSession/Cookie.php b/src/SlimSession/Cookie.php
index 911bfe9..f322b45 100644
--- a/src/SlimSession/Cookie.php
+++ b/src/SlimSession/Cookie.php
@@ -36,7 +36,10 @@ class Cookie
         ];
 
         $new = PHP_VERSION_ID >= 70300;
-        $samesite = $params['samesite'];
+        $samesite = '';
+        if (array_key_exists('samesite', $params)) {
+            $samesite = $params['samesite'];
+        }
         if ($new) {
             $args['samesite'] = $samesite;
             if ($set) {

```

**Comment by bryanjhv** [comment]
The only place I would use `array_key_exists` is when a `null` value needs to be honored, this is not the case.
Using `@` is only a bad practice if abused. In this case, it's use is controlled and predictable. No boilerplate required.

**Comment by tersmitten** [comment]
Ok, fine by me.

**Comment by tersmitten** [comment]
`!empty` or `!isset` are also options, but indeed require 4 lines instead of 1.

**Comment by tersmitten** [comment]
![Screenshot from 2021-07-02 18-13-57](https://user-images.githubusercontent.com/3392962/124302377-6e134780-db61-11eb-9371-fe9c54954279.png)

:-)

**Comment by tersmitten** [comment]
Do you want me to make 3 PR's or do it yourself?

**Comment by bryanjhv** [comment]
```php
// 2.x (PHP 5.3+), 3.x (PHP 5.4+)
$samesite = isset($params['samesite']) ? $params['samesite'] : '';
// 4.x (PHP 7.1+)
$samesite = $params['samesite'] ?? '';
```

The operator was introduced as syntactic sugar as per the docs (equivalent snippets above).
https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op
So they also want to reduce code, so it's more readable I guess. I like to keep it that way.

> Do you want me to make 3 PR's or do it yourself?

If you feel ok by committing code using `@`, you can PR the 3 branches, would help me a lot.

**Comment by tersmitten** [comment]
Sure, no problem

**Comment by tersmitten** [comment]
Fixed in #61, #62 and #63

**Comment by bryanjhv** [comment]
Thank you for contributing!
The three branches are now released:

2.3.1, 3.7.1, 4.1.1
